src/link/script.y

```diff
@@ -0,0 +1,479 @@
+	if (uint16_t offset = pc - typeInfo.startAddr; length > typeInfo.size - offset) {
+		scriptError(context, "Cannot align: the next suitable address after $%04" PRIx16 " is $%04" PRIx16 ", past $%04" PRIx16,
+		            pc, (uint16_t)(pc + length), (uint16_t)(endaddr(activeType) + 1));
+	} else {
+		pc += length;
+	}
+}
+
+static void pad(uint32_t length) {
+	auto const &context = lexerStack.back();
+	auto const &typeInfo = sectionTypeInfo[activeType];
+	auto &pc = curAddr[activeType][activeBankIdx];
+
+	assert(pc >= typeInfo.startAddr);
+	if (uint16_t offset = pc - typeInfo.startAddr; length + offset > typeInfo.size) {
+		scriptError(context, "Cannot pad by %u bytes: only %u bytes to $%04" PRIx16,
+		            length, typeInfo.size - offset, (uint16_t)(endaddr(activeType) + 1));
+	} else {
+		pc += length;
+	}
+}
+
+static void placeSection(std::string const &name) {
+	auto const &context = lexerStack.back();
+	auto const &typeInfo = sectionTypeInfo[activeType];
+
+	// A type *must* be active.
+	if (activeType == SECTTYPE_INVALID) {
+		scriptError(context, "No memory region has been specified to place section \"%s\" in",
+		            name.c_str());
+		return;
+	}
+
+	auto *section = sect_GetSection(name.c_str());
+	if (!section) {
+		scriptError(context, "Unknown section \"%s\"", name.c_str());
+		return;
+	}
+
+	assert(section->offset == 0);
+	// Check that the linker script doesn't contradict what the code says.
+	if (section->type == SECTTYPE_INVALID) {
+		// SDCC areas don't have a type assigned yet, so the linker script is used to give them one.
+		for (Section *fragment = section; fragment; fragment = fragment->nextu) {
+			fragment->type = activeType;
+		}
+	} else if (section->type != activeType) {
+		scriptError(context, "\"%s\" is specified to be a %s section, but it is already a %s section",
+		            name.c_str(), typeInfo.name.c_str(), sectionTypeInfo[section->type].name.c_str());
+	}
+
+	uint32_t bank = activeBankIdx + typeInfo.firstBank;
+	if (section->isBankFixed && bank != section->bank) {
+		scriptError(context, "Linker script wants section \"%s\" to go to %s bank %" PRIu32 ", but it is already in bank %" PRIu32,
+		            name.c_str(), sectionTypeInfo[section->type].name.c_str(), bank, section->bank);
+	}
+	section->isBankFixed = true;
+	section->bank = bank;
+
+	uint16_t &org = curAddr[activeType][activeBankIdx];
+	if (section->isAddressFixed && org != section->org) {
+		scriptError(context, "Linker script wants section \"%s\" to go to address $%04" PRIx16 ", but it is already at $%04" PRIx16,
+		            name.c_str(), org, section->org);
+	} else if (section->isAlignFixed && (org & section->alignMask) != section->alignOfs) {
+		uint8_t alignment = std::countr_one(section->alignMask);
+		scriptError(context, "Linker script wants section \"%s\" to go to address $%04" PRIx16 ", but that would be ALIGN[%" PRIu8 ", %" PRIu16 "] instead of the ALIGN[%" PRIu8 ", %" PRIu16 "] it requests",
```

Comment:
```suggestion
		scriptError(context, "\"%s\" is specified to be at address $%04" PRIx16 ", but that would be ALIGN[%" PRIu8 ", %" PRIu16 "] instead of its defined ALIGN[%" PRIu8 ", %" PRIu16 "]",
```